### PR TITLE
fix: Fixes for mcore distributed optimizer

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [ ] Does the trainer resume and restore model state all states?
- [ ] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [ ] Does the trainer support `max_steps=-1` and `validation`?
- [ ] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [ ] Does the trainer have proper logging?

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo_aligner/utils/utils.py [hunk 1]
```diff
@@ -500,6 +501,11 @@ def retrieve_model_state_dict_in_cpu(model, megatron_amp_O2=True):
     """
     cpu_dict = {}
 
+    if hasattr(model, "model"):
+        model = model.model
+    if isinstance(model, list):
+        model = model[0]
```
Comment (terrykong) [inline nemo_aligner/utils/utils.py:507]: should this assert that the length is 1? might be surprising behavior to drop the other model

File: nemo_aligner/utils/train_utils.py [hunk 2]
```diff
@@ -77,6 +77,10 @@ def prepare_for_training_step(ptl_model, zero_grad=True):
         if ptl_model.prev_global_batch_size != current_global_batch_size and ptl_model.prev_global_batch_size:
             ptl_model.trainer.should_stop = True
 
+    if ptl_model.use_mcore_dist_optim:
```
Comment (terrykong) [inline nemo_aligner/utils/train_utils.py:80]: does this need to be moved under the `if zero_grad:` line?

File: nemo_aligner/utils/utils.py [hunk 3]
```diff
@@ -552,14 +558,19 @@ def swap_dict(resident_model, cpu_weights, offload_onto_cpu=True, megatron_amp_O
     if offload_onto_cpu:
         offloaded_weights = retrieve_model_state_dict_in_cpu(resident_model, megatron_amp_O2=megatron_amp_O2)
 
-    resident_model.load_state_dict(cpu_weights)
+    load_to = resident_model.model
+    if isinstance(load_to, list):
+        load_to = load_to[0]
```
Comment (terrykong) [inline nemo_aligner/utils/utils.py:563]: same Q: about the length

for my understanding, is this a change needed b/c of mcore dist opt?

File: nemo_aligner/utils/train_utils.py (hunk 2 above)
Comment (ashors1) [reply nemo_aligner/utils/train_utils.py:80]: This needs to be done regardless of whether `zero_grad` is true. The reason `zero_grad` is sometimes false is because zeroing the gradient is handled from the custom trainer. But I thought it made more sense to zero the grad buffers here to avoid duplicating that code across all trainers.

File: nemo_aligner/utils/utils.py (hunk 3 above)
Comment (ashors1) [reply nemo_aligner/utils/utils.py:563]: Yeah, when using mcore optimizer, the model ends up being a list of McoreDDP-wrapped model chunks (references: https://github.com/NVIDIA/NeMo/blob/0dd33007e6e37bc8fb5bff152595e46a82eccdae/nemo/collections/nlp/parts/nlp_overrides.py#L282-L283, https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/nlp/models/language_modeling/megatron_gpt_model.py#L609-L620). This isn't the case when using apex.

File: nemo_aligner/utils/utils.py [hunk 4]
```diff
@@ -552,14 +561,22 @@ def swap_dict(resident_model, cpu_weights, offload_onto_cpu=True, megatron_amp_O
     if offload_onto_cpu:
         offloaded_weights = retrieve_model_state_dict_in_cpu(resident_model, megatron_amp_O2=megatron_amp_O2)
 
-    resident_model.load_state_dict(cpu_weights)
+    load_to = resident_model.model
+    if isinstance(load_to, list):
+        assert (
+            len(model) == 1
```
[the remark this answers is not in the document]
Comment (ashors1) [reply nemo_aligner/utils/utils.py]: Oh yes, thanks for the catch!

**Comment by terrykong** [comment]
Latest CI passed: https://github.com/NVIDIA/NeMo-Aligner/actions/runs/13416682771

recent CI failure due to cherry-picking label
